File: src/com/mantono/webserver/RequestResponder.java
package com.mantono.webserver;

import com.mantono.webserver.rest.HeaderField;
import com.mantono.webserver.rest.Resource;
import com.mantono.webserver.rest.Response;
import com.mantono.webserver.rest.ResponseCode;

import java.io.IOException;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.net.Socket;
import java.net.URISyntaxException;
import java.time.Duration;
import java.time.Instant;
import java.util.Map;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.TimeUnit;

public class RequestResponder implements Runnable
{
	private final Map<Resource, Method> resourceMap;
	private final BlockingQueue<Socket> incoming;
	
	public RequestResponder(final Map<Resource, Method> resources, final BlockingQueue<Socket> requests)
	{
		this.resourceMap = resources;
		this.incoming = requests;
	}

	@Override
	public void run()
	{
		while(true)
		{
			try(final Socket socket = incoming.poll(30, TimeUnit.SECONDS))
			{
				final Instant start = Instant.now();
				if(socket == null)
					continue;

				RequestParser rParser = new RequestParser(socket, resourceMap);
				Request request = rParser.getRequest();

				Response response = execute(rParser.getMethod(), request);
				ResponseSender sender = new ResponseSender(socket, response);
				sender.send();
				sender.close();
				final Instant end = Instant.now();
				final Duration duration = Duration.between(start, end);
				System.out.println(duration);
			}
			catch(InterruptedException e)
			{
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
			catch(IOException e)
			{
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
			catch(URISyntaxException e)
			{
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
			catch(IllegalAccessException e)
			{
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
			catch(IllegalArgumentException e)
			{
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
			catch(InvocationTargetException e)
			{
				// TODO Auto-generated catch block
				e.printStackTrace();
			}
		}
	}

	private void notFound(final Socket socket) throws IOException
	{
		WebPage response = new WebPage(ResponseCode.NOT_FOUND);
		response.getHeader().set(HeaderField.CONTENT_LENGTH, "0");
		ResponseSender rs = new ResponseSender(socket, response);
		rs.send();
		rs.close();
	}

	private Response execute(Method method, Request request) throws IllegalAccessException, IllegalArgumentException, InvocationTargetException
	{
		Class<?>[] parameterTypes = method.getParameterTypes();

		if(parameterTypes.length == 0)
			return (Response) method.invoke(null);
		if(parameterTypes.length > 1)
			throw new IllegalArgumentException("Found method " + method + " that has parameter than one parameter");
		if(parameterTypes[0] != Request.class)
			throw new IllegalArgumentException("Found method " + method + " that has parameter that is not a Request ("+parameterTypes[0]+")");

		return (Response) method.invoke(null, request);
	}
}

File: src/main/java/com/mantono/webserver/RequestResponder.kt
package com.mantono.webserver

import com.mantono.webserver.reflection.staticResources
import com.mantono.webserver.rest.*
import java.io.File
import java.io.IOException
import java.io.InputStream
import java.lang.reflect.InvocationTargetException
import java.lang.reflect.Method
import java.net.Socket
import java.nio.file.Files
import java.nio.file.Path
import java.time.Duration
import java.time.Instant
import java.util.concurrent.BlockingQueue

class RequestResponder(private val resourceMap: Map<Resource, Method>, private val incoming: BlockingQueue<Socket>) : Runnable
{

	override fun run()
	{
		while (true)
		{
			try
			{
				incoming.take().use {socket ->
					val start = Instant.now()
					println("Got a connection!")

					val request = parseRequest(socket, resourceMap)
					val response: Response = when (request)
					{
						is ValidRequest -> createResponse(resourceMap, request)
						is ValidStaticRequest -> createStaticResponse(request)
						is InvalidRequest -> notFound(request)
					}

					send(socket, response)
					val end = Instant.now()
					val duration = Duration.between(start, end)
					println(duration)
				}
			}
			catch (e: InterruptedException)
			{
				// TODO Auto-generated catch block
				e.printStackTrace()
			}
			catch (e: IOException)
			{
				// TODO Auto-generated catch block
				e.printStackTrace()
			}
			catch (e: IllegalAccessException)
			{
				// TODO Auto-generated catch block
				e.printStackTrace()
			}
			catch (e: IllegalArgumentException)
			{
				// TODO Auto-generated catch block
				e.printStackTrace()
			}
			catch (e: InvocationTargetException)
			{
				// TODO Auto-generated catch block
				e.printStackTrace()
			}

		}
	}

	private fun createResponse(resourceMap: Map<Resource, Method>, request: ValidRequest): Response
	{
		val method = resourceMap[request.resource]!!
		return execute(method, request)
	}

	private fun  createStaticResponse(request: ValidStaticRequest): Response
	{
		val loadedFile: CharSequence = loadFile(request.file)
		val fields = mapOf(HeaderField.CONTENT_TYPE to "image/png")
		val header: ResponseHeader = ResponseHeader(fields)
		return SimpleResponse(ResponseCode.OK, header, body = loadedFile)
	}

	private fun loadFile(filePath: Path): CharSequence
	{
		if(!filePath.toFile().exists())
			throw IllegalStateException("File does not exist: $filePath")
		return String(Files.readAllBytes(filePath))
	}

	private fun notFound(request: InvalidRequest): Response
	{
		// TODO Check if there is a default method for invalid request not found
		// should be annotated with @NotFound (to be implemented), else return this response
		return SimpleResponse(ResponseCode.NOT_FOUND)
	}

	private fun execute(method: Method, request: Request): Response
	{
		val parameterTypes = method.parameterTypes

		if (parameterTypes.size == 0)
			return method.invoke(null) as Response
		return method.invoke(null, request) as Response
	}
}